### PR TITLE
Print requires_grad and grad_fn in string repr of tensor

### PR DESCRIPTION
**Pull request by colesbury**
```
For example:

  >>> torch.ones(3).requires_grad_()
  tensor([ 1.,  1.,  1.], requires_grad=True)

  >>> torch.ones(3).requires_grad_() * 5
  tensor([ 5.,  5.,  5.], grad_fn=<MulBackward0>)

The suffix (dtype, requires_grad, grad_fn) wraps to a new line if
it would cause the the line to exceed the linewidth.

  >>> torch.ones(10).double().requires_grad_()
  tensor([ 1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.],
         dtype=torch.float64, requires_grad=True)
```

### CODE REVIEW

**Comment by soumith** [comment]
cc: @li-roy your PR https://github.com/pytorch/pytorch/pull/7632 might conflict with this ?

**Comment by apaszke** [comment]
Do we really want to show implementation details like `MulBackward0` as part of our public interface? What's a benefit of seeing the grad function?

**Comment by colesbury** [comment]
Regarding `grad_fn`:

```torch.ones(10).double().requires_grad_()```

and

```torch.ones(10).requires_grad_().double()```

are very different. Make this distinction visible is important for debugging.

**Comment by colesbury** [comment]
To expand on the above point: a tensor with history is behaves differently from a (leaf) tensor that has requires_grad=True. This includes whether the `.grad` is stored during `backward()` and whether the tensor may be holding on to a large computational graph.

This is a common source of confusion, especially for people new to PyTorch. We should ease this confusion by making it immediately visible when a tensor has history (a grad_fn). Printing out the name of the grad_fn gives more useful information than something like `is_leaf=False`.

**Comment by apaszke** [comment]
I would be ok with printing `is_leaf=False`, but I think `grad_fn` is too much of an implementation detail. It's not documented, it's not guaranteed to remain the same, it's not even very readable unless you know how autograd works. People should never depend on how the graph is constructed.

**Comment by colesbury** [comment]
We mention grad_fn in the docs (but not `is_leaf`). I agree that the name backward function is an implementation detail, but it's often useful to print out implementation details. You see this often in Python where:

```
>>> Foo()
<Foo at 0x7fa5678cf160>
>>> socket.socket()
<socket.socket fd=13, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('0.0.0.0', 0)>
```

The socket's file descriptor and object's memory address are clearly implementation details, but they help in debugging.

The name of grad_fn helps you understand what operation created the variable. The names aren't perfect, but they're usually enough.

**Comment by apaszke** [comment]
This is different. The `socket` module is pretty much a direct wrapper around the C API, and you really need to understand pretty much most of the ops it prints to do anything with sockets. Autograd is different. Most users don't really know or even should now how it works under the hood, and directly showing `grad_fn` is something that is very internal. If the problem is that "leaf" is not a widely recognized name, can we make it `has_grad_fn=True`?

**Comment by apaszke** [comment]
Another downside is that `__repr__` no longer gives you a constructor call you could use.

**Comment by soumith** [comment]
I'm with Sam on this one. It seems really useful to glance at grad_fn in the repr, and it's not like a strict design pattern in Python-land for repr. How many times do we actually use the "repr is constructor" behavior, almost never.

**Comment by gchanan** [comment]
I don't feel super strong about this either way, but we could include the grad_fn in a comment on the next line, so that we retain the "repr is constructor" behavior while still being informative.

**Comment by colesbury** [comment]
The advantage of the current repr style syntax is that it's immediately familiar and understood to most Python users, not so much the copy-paste ability.

The literal copy-pasting to recreate the object tends not to work in practice because of summarizing of large tensors, floating point truncation, and because most people don't have `tensor()` directly imported. Furthermore, objects with grad_fn are not value objects -- they have identity. Python often uses `<Foo ...>` syntax for non-value objects, but I don't think it makes sense to do that here.

`has_grad_fn=True` doesn't do enough to convey the importance of `grad_fn`. It looks like a simple boolean attribute. `grad_fn=<MulBackward>` looks like an object, which is important because it may hold on to large amounts of state. We've clearly thought that this structure is important for users to understand, since we've put it on the [front page of our GitHub](https://github.com/pytorch/pytorch/raw/master/docs/source/_static/img/dynamic_graph.gif) since release.

**Comment by apaszke** [comment]
I still think it adds a lot of unnecessary noise to the repr, but if this is really what everyone else wants then let's add it.

**Review by soumith — APPROVED** [review]
